File: placement-flutter/storyly_placement_flutter/ios/Classes/StorylyPlacementFlutterPlugin.swift
import Flutter
import UIKit

public class StorylyPlacementFlutterPlugin: NSObject, FlutterPlugin {
  private let channel: FlutterMethodChannel

  init(channel: FlutterMethodChannel) {
    self.channel = channel
  }

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "storyly_placement_flutter", binaryMessenger: registrar.messenger())
    let instance = StorylyPlacementFlutterPlugin(channel: channel)
    registrar.addMethodCallDelegate(instance, channel: channel)
      
    let factory = StorylyPlacementFlutterViewFactory(messenger: registrar.messenger())
    registrar.register(factory, withId: "storyly_placement_flutter_view")
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getPlatformVersion":
      result("iOS " + UIDevice.current.systemVersion)
    case "createProvider":
        if let providerId = call.arguments as? String {
            let wrapper = SPPlacementProviderManager.shared.createProvider(id: providerId)
            wrapper.sendEvent = { [weak self] id, event, eventData in
                DispatchQueue.main.async {
                    let payload: [String: Any] = [
                        "providerId": id,
                        "raw": eventData
                    ]
                    self?.channel.invokeMethod(event.eventName, arguments: payload)
                }
            }
            result(nil)
        } else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "providerId is required", details: nil))
        }
    case "destroyProvider":
        if let providerId = call.arguments as? String {
            SPPlacementProviderManager.shared.destroyProvider(id: providerId)
            result(nil)
        } else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "providerId is required", details: nil))
        }
    case "updateConfig":
        if let args = call.arguments as? [String: Any],
           let providerId = args["providerId"] as? String,
           let config = args["config"] as? String {
            SPPlacementProviderManager.shared.getProvider(id: providerId)?.configure(configJson: config)
            result(nil)
        } else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "providerId and config are required", details: nil))
        }
    case "hydrateProducts":
        if let args = call.arguments as? [String: Any],
           let providerId = args["providerId"] as? String,
           let products = args["products"] as? String {
            SPPlacementProviderManager.shared.getProvider(id: providerId)?.hydrateProducts(productsJson: products)
            result(nil)
        } else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "providerId and products are required", details: nil))
        }
    case "hydrateWishlist":
        if let args = call.arguments as? [String: Any],
           let providerId = args["providerId"] as? String,
           let products = args["products"] as? String {
            SPPlacementProviderManager.shared.getProvider(id: providerId)?.hydrateWishlist(productsJson: products)
            result(nil)
        } else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "providerId and products are required", details: nil))
        }
    case "updateCart":
        if let args = call.arguments as? [String: Any],
           let providerId = args["providerId"] as? String,
           let cart = args["cart"] as? String {
            SPPlacementProviderManager.shared.getProvider(id: providerId)?.updateCart(cartJson: cart)
            result(nil)
        } else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "providerId and cart are required", details: nil))
        }
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}

class StorylyPlacementFlutterViewFactory: NSObject, FlutterPlatformViewFactory {
    private var messenger: FlutterBinaryMessenger

    init(messenger: FlutterBinaryMessenger) {
        self.messenger = messenger
        super.init()
    }

    func create(
        withFrame frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?
    ) -> FlutterPlatformView {
        return StorylyPlacementFlutterView(
            frame: frame,
            viewIdentifier: viewId,
            arguments: args,
            messenger: messenger
        )
    }
    
    public func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
          return FlutterStandardMessageCodec.sharedInstance()
    }
}

class StorylyPlacementFlutterView: NSObject, FlutterPlatformView {
    private var _view: UIView
    private var placementView: SPStorylyPlacementView
    private var methodChannel: FlutterMethodChannel

    init(
        frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?,
        messenger: FlutterBinaryMessenger
    ) {
        _view = UIView()
        placementView = SPStorylyPlacementView()
        
        methodChannel = FlutterMethodChannel(name: "storyly_placement_flutter/view_\(viewId)", binaryMessenger: messenger)
        
        super.init()
        
        if let args = args as? [String: Any],
           let providerId = args["providerId"] as? String {
            placementView.configure(providerId: providerId)
        }
        
        setupMethodChannel()
        setupCallbacks()
        
        placementView.frame = frame
        placementView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        _view.addSubview(placementView)
    }

    func view() -> UIView {
        return _view
    }
    
    private func setupMethodChannel() {
        methodChannel.setMethodCallHandler { [weak self] (call: FlutterMethodCall, result: @escaping FlutterResult) in
            guard let self = self else { return }
            
            switch call.method {
            case "configure":
                if let args = call.arguments as? [String: Any],
                   let providerId = args["providerId"] as? String {
                    self.placementView.configure(providerId: providerId)
                    result(nil)
                } else {
                    result(FlutterError(code: "INVALID_ARGUMENT", message: "providerId is required", details: nil))
                }
            case "callWidget":
                 if let args = call.arguments as? [String: Any],
                    let viewId = args["viewId"] as? String,
                    let method = args["method"] as? String,
                    let raw = args["raw"] as? String {
                     self.placementView.callWidget(id: viewId, method: method, raw: raw)
                     result(nil)
                 } else {
                     result(FlutterError(code: "INVALID_ARGUMENT", message: "viewId, method and params are required", details: nil))
                 }
            case "approveCartChange":
                if let args = call.arguments as? [String: Any],
                   let responseId = args["responseId"] as? String,
                   let raw = args["raw"] as? String {
                    self.placementView.approveCartChange(responseId: responseId, raw: raw)
                    result(nil)
                } else {
                    result(FlutterError(code: "INVALID_ARGUMENT", message: "responseId and cart are required", details: nil))
                }
            case "rejectCartChange":
                if let args = call.arguments as? [String: Any],
                   let responseId = args["responseId"] as? String,
                   let raw = args["raw"] as? String {
                    self.placementView.rejectCartChange(responseId: responseId, raw: raw)
                    result(nil)
                } else {
                    result(FlutterError(code: "INVALID_ARGUMENT", message: "responseId and failMessage are required", details: nil))
                }
            case "approveWishlistChange":
                if let args = call.arguments as? [String: Any],
                   let responseId = args["responseId"] as? String,
                   let raw = args["raw"] as? String {
                    self.placementView.approveWishlistChange(responseId: responseId, raw: raw)
                    result(nil)
                } else {
                    result(FlutterError(code: "INVALID_ARGUMENT", message: "responseId and item are required", details: nil))
                }
            case "rejectWishlistChange":
                if let args = call.arguments as? [String: Any],
                   let responseId = args["responseId"] as? String,
                   let raw = args["failMessage"] as? String {
                    self.placementView.rejectWishlistChange(responseId: responseId, raw: raw)
                    result(nil)
                } else {
                    result(FlutterError(code: "INVALID_ARGUMENT", message: "responseId and failMessage are required", details: nil))
                }
            default:
                result(FlutterMethodNotImplemented)
            }
        }
    }
    
    private func setupCallbacks() {
        placementView.dispatchEvent = { [weak self] event, eventData in
            self?.methodChannel.invokeMethod(event.eventName, arguments: eventData)
        }
    }
}

File: flutter/storyly_placement_flutter/android/src/main/kotlin/com/appsamurai/storyly/storyly_placement_flutter/StorylyPlacementFlutterPlugin.kt
package com.appsamurai.storyly.storyly_placement_flutter

import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import com.appsamurai.storyly.storyly_placement_flutter.common.SPPlacementProviderManager
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import com.appsamurai.storyly.storyly_placement_flutter.common.data.util.decodeFromJson

/** StorylyPlacementFlutterPlugin */
class StorylyPlacementFlutterPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var applicationContext: Context

    private var placementFlutterViewFactory: StorylyPlacementFlutterViewFactory? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "storyly_placement_flutter")
        channel.setMethodCallHandler(this)


        placementFlutterViewFactory = StorylyPlacementFlutterViewFactory(flutterPluginBinding.binaryMessenger).also { factory ->
            flutterPluginBinding.platformViewRegistry.registerViewFactory(
                "storyly_placement_flutter_view",
                factory,
            )
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(
        call: MethodCall,
        result: Result
    ) {
        when (call.method) {
            "getPlatformVersion" -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }
            "createProvider" -> {
                val providerId = call.arguments as? String
                if (providerId != null) {
                    SPPlacementProviderManager.createProvider(applicationContext, providerId).apply {
                        this.sendEvent = { id, event, eventData ->
                            Handler(Looper.getMainLooper()).post {
                                val payload = mapOf(
                                    "providerId" to id,
                                    "raw" to eventData
                                )
                                channel.invokeMethod(event.eventName, payload)
                            }
                        }
                    }
                    result.success(null)
                } else {
                    result.error("INVALID_ARGUMENT", "providerId is required", null)
                }
            }
            "destroyProvider" -> {
                val providerId = call.arguments as? String
                if (providerId != null) {
                    SPPlacementProviderManager.destroyProvider(providerId)
                    result.success(null)
                } else {
                    result.error("INVALID_ARGUMENT", "providerId is required", null)
                }
            }
            "updateConfig" -> {
                val args = call.arguments as? Map<String, Any>
                val providerId = args?.get("providerId") as? String ?: return
                val config = args["config"] as? String

                val provider = SPPlacementProviderManager.getProvider(providerId) ?: return
                
                if (config != null) {
                    provider.configure(config)
                    result.success(null)
                } else {
                    result.error("INVALID_ARGUMENT", "providerId and config are required", null)
                }
            }
            "hydrateProducts" -> {
                val args = call.arguments as? Map<String, Any>
                val providerId = args?.get("providerId") as? String ?: return
                val products = args["products"] as? String

                val provider = SPPlacementProviderManager.getProvider(providerId) ?: return
                
                if (products != null) {
                    provider.hydrateProducts(products)
                    result.success(null)
                } else {
                    result.error("INVALID_ARGUMENT", "providerId and products are required", null)
                }
            }
            "hydrateWishlist" -> {
                val args = call.arguments as? Map<String, Any>
                val providerId = args?.get("providerId") as? String ?: return
                val products = args["products"] as? String

                val provider = SPPlacementProviderManager.getProvider(providerId) ?: return
                
                if (products != null) {
                    provider.hydrateWishlist(products)
                    result.success(null)
                } else {
                    result.error("INVALID_ARGUMENT", "providerId and products are required", null)
                }
            }
            else -> {
                result.notImplemented()
            }
        }
    }
}
